language/stdlib/modules/designated_dealer.move

```diff
@@ -0,0 +1,129 @@
+        /// 0-indexed array of tier upperbounds
+        tiers: vector<u64>
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    // To-be designated-dealer called functions
+    ///////////////////////////////////////////////////////////////////////////
+
+    public fun create_designated_dealer(
+    ): Dealer {
+        Dealer {
+            window_start: LibraTimestamp::now_microseconds(),
+            window_inflow: 0,
+            is_certified: true,
+            tiers: Vector::empty(),
+        }
+    }
+
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Publicly callable APIs by Treasury Compliance Account
+    ///////////////////////////////////////////////////////////////////////////
+
+    public fun add_tier(dealer: &mut Dealer, next_tier_upperbound: u64)
+    {
+        assert_sender_is_blessed();
+        let tiers = &mut dealer.tiers;
+        let number_of_tiers: u64 = Vector::length(tiers);
+        // INVALID_TIER_ADDITION
+        Txn::assert(number_of_tiers <= 4, 3);
+        if (number_of_tiers > 1) {
+            let prev_tier = *Vector::borrow(tiers, number_of_tiers - 1);
+            // INVALID_TIER_START
+            Txn::assert(prev_tier < next_tier_upperbound, 4);
+        };
+        Vector::push_back(tiers, next_tier_upperbound);
+    }
+
+    public fun update_tier(dealer: &mut Dealer, tier_index: u64, new_upperbound: u64)
+    {
+        assert_sender_is_blessed();
+        let tiers = &mut dealer.tiers;
+        let number_of_tiers = Vector::length(tiers);
+        // INVALID_TIER_INDEX
+        Txn::assert(tier_index <= 4, 3);
+        Txn::assert(tier_index < number_of_tiers, 3);
+        // Make sure that this new start for the tier is consistent
+        // with the tiers above it.
+        let index = tier_index + 1;
+        while (index < number_of_tiers) {
+            // INVALID_TIER_START
+            Txn::assert(new_upperbound < *Vector::borrow(tiers, index), 4);
+            index = index + 1;
+        };
+        let tier_mut = Vector::borrow_mut(tiers, tier_index);
+        *tier_mut = new_upperbound;
+    }
+
+    public fun tiered_mint(dealer: &mut Dealer, amount: u64, tier_index: u64): bool {
+        reset_window(dealer);
+        let cur_inflow = *&dealer.window_inflow;
+        let tiers = &mut dealer.tiers;
+        // If the tier_index is one past the bounded tiers, minting is unbounded
+        let number_of_tiers = Vector::length(tiers);
+        let tier_check = &mut false;
+        if (tier_index == number_of_tiers) {
+            *tier_check = true;
+        } else {
+            let tier_upperbound: u64 = *Vector::borrow(tiers, tier_index);
+            *tier_check = (cur_inflow + amount <= tier_upperbound);
+        };
+        if (*tier_check) {
+            dealer.window_inflow = cur_inflow + amount;
+        };
+        *tier_check
+    }
+
+    public fun remove_designated_dealer(dealer: &mut Dealer)
```

Comment:
Wondering: what should a DD with `is_certified = false` be able to do? Would it be better/simpler to just freeze the account?